File: apps/musician/lib/services/chat_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import '../models/chat_model.dart';

class ChatService extends ChangeNotifier {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  String? get currentUserId => _auth.currentUser?.uid;

  Stream<List<ChatModel>> getChats() {
    final uid = currentUserId;
    if (uid == null) return Stream.value([]);

    return _firestore
        .collection('chats')
        .where('participantIds', arrayContains: uid)
        .orderBy('lastMessageTime', descending: true)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => ChatModel.fromFirestore(doc)).toList();
    });
  }

  Stream<List<MessageModel>> getMessages(String chatId) {
    return _firestore
        .collection('chats')
        .doc(chatId)
        .collection('messages')
        .orderBy('timestamp', descending: false)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => MessageModel.fromFirestore(doc)).toList();
    });
  }

  Future<void> sendMessage(String chatId, String text) async {
    final uid = currentUserId;
    if (uid == null) return;

    final messageData = {
      'senderId': uid,
      'text': text,
      'timestamp': FieldValue.serverTimestamp(),
      'type': 'text',
    };

    final batch = _firestore.batch();
    
    // Add message
    final messageRef = _firestore
        .collection('chats')
        .doc(chatId)
        .collection('messages')
        .doc();
    batch.set(messageRef, messageData);

    // Update chat last message
    final chatRef = _firestore.collection('chats').doc(chatId);
    batch.update(chatRef, {
      'lastMessage': text,
      'lastMessageTime': FieldValue.serverTimestamp(),
      'lastMessageSenderId': uid,
    });

    await batch.commit();
  }

  Future<String> getOrCreateChat(String otherUserId, String otherUserName, String otherUserImage) async {
    final uid = currentUserId;
    if (uid == null) throw Exception('User not authenticated');

    // Always fetch the organizer's real name first
    String resolvedName = otherUserName;
    String resolvedImage = otherUserImage;
    try {
      final otherDoc = await _firestore.collection('organizers').doc(otherUserId).get();
      if (otherDoc.exists) {
        resolvedName = otherDoc.data()?['name'] ?? otherUserName;
        resolvedImage = otherDoc.data()?['profileImageUrl'] ?? otherUserImage;
      }
    } catch (e) {
      debugPrint('Could not fetch organizer profile for chat: $e');
    }

    // Get current musician's real name
    final currentUserDoc = await _firestore.collection('musicians').doc(uid).get();
    final currentUserName = currentUserDoc.data()?['fullName'] ?? 'Musician';
    final currentUserImage = currentUserDoc.data()?['profileImageUrl'] ?? '';

    // Check if chat already exists — if so, patch names and return
    final existingChat = await _firestore
        .collection('chats')
        .where('participantIds', arrayContains: uid)
        .get();

    for (var doc in existingChat.docs) {
      final List<String> participants = List<String>.from(doc['participantIds']);
      if (participants.contains(otherUserId)) {
        // Patch with correct names/images in case they were wrong before
        await doc.reference.update({
          'participantNames.$uid': currentUserName,
          'participantNames.$otherUserId': resolvedName,
          'participantImages.$uid': currentUserImage,
          'participantImages.$otherUserId': resolvedImage,
        });
        return doc.id;
      }
    }

    // Create new chat with correct data
    final newChatRef = _firestore.collection('chats').doc();
    await newChatRef.set({
      'participantIds': [uid, otherUserId],
      'participantNames': {
        uid: currentUserName,
        otherUserId: resolvedName,
      },
      'participantImages': {
        uid: currentUserImage,
        otherUserId: resolvedImage,
      },
      'lastMessage': '',
      'lastMessageTime': FieldValue.serverTimestamp(),
      'lastMessageSenderId': '',
      'unreadCount': {
        uid: 0,
        otherUserId: 0,
      },
    });

    return newChatRef.id;
  }
}

File: backend/services/chat_service.py
from firebase_admin import firestore
from backend.database import db
from typing import List, Optional, Dict

class ChatService:
    @staticmethod
    def get_or_create_chat(participant_ids: List[str], participant_names: Dict[str, str], participant_images: Dict[str, str]):
        # Sort participant IDs to ensure a consistent chat ID between two users
        participant_ids.sort()
        
        # Check if chat already exists
        chats = db.collection("chats")\
                  .where("participantIds", "==", participant_ids)\
                  .limit(1)\
                  .get()
        
        if len(chats) > 0:
            return chats[0].id
            
        # Create new chat
        chat_data = {
            "participantIds": participant_ids,
            "participantNames": participant_names,
            "participantImages": participant_images,
            "lastMessage": "Chat started",
            "lastMessageTime": firestore.SERVER_TIMESTAMP,
            "createdAt": firestore.SERVER_TIMESTAMP,
            "unreadCount": {p_id: 0 for p_id in participant_ids}
        }
        
        doc_ref = db.collection("chats").document()
        doc_ref.set(chat_data)
        return doc_ref.id
        
    @staticmethod
    def send_message(chat_id: str, sender_id: str, text: str):
        from backend.services.notification_service import NotificationService
        
        chat_ref = db.collection("chats").document(chat_id)
        chat_doc = chat_ref.get()
        if not chat_doc.exists:
            return None
            
        chat_data = chat_doc.to_dict()
        participant_ids = chat_data.get("participantIds", [])
        recipient_id = next((p_id for p_id in participant_ids if p_id != sender_id), None)
        
        # 1. Add message to subcollection
        message_data = {
            "senderId": sender_id,
            "text": text,
            "timestamp": firestore.SERVER_TIMESTAMP,
            "type": "text"
        }
        chat_ref.collection("messages").add(message_data)
        
        # 2. Update chat metadata
        chat_ref.update({
            "lastMessage": text,
            "lastMessageTime": firestore.SERVER_TIMESTAMP,
            "lastMessageSenderId": sender_id
        })
        
        # 3. Trigger Push Notification to recipient
        if recipient_id:
            sender_name = chat_data.get("participantNames", {}).get(sender_id, "Someone")
            NotificationService.send_to_user(
                user_id=recipient_id,
                title=f"Message from {sender_name}",
                body=text,
                data={"chatId": chat_id, "type": "chat"}
            )
            
        return True
